### DISCUSSION TITLE
[FluentCalendar] Select multiple dates and range of dates

### DISCUSSION THREAD

**Discussion by dvoituron** [microsoft/fluentui-blazor#2172] [Ideas]
# [FluentCalendar] Select multiple dates and range of dates

We're thinking of adding a `SelectMode` property so that the user can select a range of dates `CalendarSelectMode.Range` or a discontinuous list of dates `CalendarSelectMode.Multiple`.

Initially, this feature will only be available for the FluentCalendar with `View=Days`.

A new bindable property `SelectedDates` will retrieve these dates, rendering the existing `Value` property useless in these `SelectModes` only.

Added a `SelectDatesHover` method that returns a list of dates to be pre-selected (hover) and activated when the user clicks on this selection. This allows you to manage choices such as selecting an entire week, with a single mouse click. Or selecting 3 days automatically, with a single click.

The `@bind-SelectedDates:after` command can be used to validate the content of selected dates, block or remove them if necessary, or even display an error message.

## Example

```razor
<FluentCalendar @bind-SelectedDates="@SelectedRangeCustomized"
                SelectDatesHover="@SelectOneWeek"
                SelectMode="CalendarSelectMode.Range" />

<FluentCalendar @bind-SelectedDates="@SelectedDaysCustomized"
                SelectDatesHover="@Select3Days"
                SelectMode="CalendarSelectMode.Multiple" />

@code
{
    private ICollection<DateTime> SelectedRangeCustomized = new List<DateTime>();
    private ICollection<DateTime> SelectedDaysCustomized = new List<DateTime>();

    private IEnumerable<DateTime> SelectOneWeek(DateTime date)
    {
        return new[] { date.StartOfWeek(), date.StartOfWeek().AddDays(6) };
    }

    private IEnumerable<DateTime> Select3Days(DateTime date)
    {
        return Enumerable.Range(0, 3).Select(i => date.AddDays(i));
    }
}
```

![DateSelection](https://github.com/microsoft/fluentui-blazor/assets/8350694/9cdc993c-6675-45f0-b3cf-9703aa21c04f)

[EDIT] 2024-06-10: Adding `SelectDatesHover` function to customize the hover effect (dates will be selected) and many samples.

**Comment by hksalessio** [comment]
I absolutely love, what this library is becoming! I hope you continue the great work!

**Comment by oneolddev** [comment]
On `SelectMode.Multiple`, it would be nice to be able to be notified each time a date is selected. The use case I'm thinking of, is for example limiting the user to selecting only 4 dates.

Looking forward to seeing it.

**Reply by dvoituron** [reply]
`SelectedDates` will be bindable. So you will have the event `SelectedDatesChanged`. But I thought to include a property like "Maximum of number dates" 😊

**Comment by dvoituron** [comment]
I added some new samples, based on a new `SelectDatesHover` method.

**Comment by Rocky-25** [comment]
It'd be good if this can be done via a date picker too so the calendar doesn't always have to be displayed and taking up space on the page.

**Reply by vnbaaij** [reply]
The date picker, by default, only allows for selecting 1 date

**Reply by Rocky-25** [reply]
The calendar can have a single date (Value) or multiple (SelectedDates) so couldn't the same functionality be done to the date picker?
Does the Date picker not use the calendar component as part of it?

**Reply by vnbaaij** [reply]
It is not about the calendar. That supports multiple dates indeed. it is about the (text field in the) picker. That one, by default, displays a single date.

**Comment by svenwanzenried** [comment]
Thanks for this feature. What i am missing is the possibility to have the mode set to "multiple" but still be able to select a range as part of that. (E.g. while pressing Shift, i can select the start and end of a range)

**Reply by dvoituron** [reply]
You already have that. Not using Shift key but selecting Range or Multiple

See examples https://fluentui-blazor-v5.azurewebsites.net/DateTime/Calendar

**Reply by svenwanzenried** [reply]
Sorry, i didn't express my use case properly. Say, I wanted to select the following days:
<img width="240" alt="image" src="https://github.com/user-attachments/assets/b3e4ba2e-3f3f-4910-989b-8b1cb26dc0d8" />

I would like to select thee ranges (5th - 10th, 14th - 17th, 26th - 27th) and the 30th, instead of clicking 13 times.
This is currently not possible, correct?

**Reply by vnbaaij** [reply]
Correct. You'll need to click 13 times indeed.

We only support selecting 1 consecutive range.
